IsolatedStorageException

OzBob opened the pull request:
method to create directory to prevent IsolatedStorageException
tabs to spaces - only to be consistent...

## Summary

Isolated storage exception example, and doc change 

Fixes #Issue_Number (if available)

OzBob committed:
IsolatedStorageException

method to create directory to prevent IsolatedStorageException
tabs to spaces - only to be consistent...

BillWagner commented on samples/snippets/csharp/VS_Snippets_CLR/conceptual.isolatedstorage/cs/source.cs:
```diff
@@ -5,21 +5,52 @@
         }
     }
+
+    private string CreateFolders(string folderName, string fileName)
```
I'm curious why you added this method @OzBob It's not referenced anywhere.

OzBob replied:
removed private function - it was there in my original draft as a more robust code example, but is not necessary for this code snippet

spacing set to 4 spaces.

BillWagner reviewed (commented):
Thanks for creating this PR @OzBob 

I had one question before we merged this.

Thraka commented:
The indentation on this has changed too, making the code file incorrectly indented.

OzBob committed:
FileName can have DirectorySeparatorChar

indenting to 4 spaces
CreateFolders function removed, originally to cater for when the filename has a Directory Separator Char in it.

OzBob committed:
indenting to 4 spaces

indenting to 4 spaces
CreateFolders function removed, originally to cater for when the filename has a Directory Separator Char in it.

OzBob committed:
indenting to 4 spaces

indenting to 4 spaces
CreateFolders function removed, originally to cater for when the filename has a Directory Separator Char in it.

OzBob committed:
indenting to 4 spaces

indenting to 4 spaces
CreateFolders function removed, originally to cater for when the filename has a Directory Separator Char in it.

adegeo reviewed (approved):
Approving. Bill reviewed and comment was addressed. PR was created before new rules related to PRs and code were created.

adegeo commented on samples/snippets/csharp/VS_Snippets_CLR/conceptual.isolatedstorage/cs/source.cs:
```diff
@@ -25,11 +25,14 @@ public static void Main()
-            isoStore.CreateDirectory("AnotherTopLevelDirectory/InsideDirectory");
-            isoStore.CreateFile("AnotherTopLevelDirectory/InsideDirectory/HereIAm.txt");
-
+            isoStore.CreateFile("AnotherTopLevelDirectory/InsideDirectory");
+            Console.WriteLine("Created directories.");
+            
```
Was this change intentional? It removes the "InsideDirectory" and then you try to create a file as a directory?

adegeo replied:
Hi @OzBob Sorry about that, I was on vacation. I just got back. What I mean is that the directory "InsideDirectory" doesn't seem to be created anymore? Unless I'm missing something, which is possible.

I just created a test project and ran the original code. It worked as expected:

```dotnet
dotnet run

Created directories.
Created a new file in the root.
Created a new file in the InsideDirectory.
```

I then ran the code from this PR and it failed:

```dotnet
dotnet run

System.IO.IsolatedStorage.IsolatedStorageException: Operation not permitted on IsolatedStorageFileStream.
 ---> System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\thrak\AppData\Local\IsolatedStorage\5uysulys.woc\5ou1xfgf.iws\Url.2txcxmkov3cbmd5tmqeaz0ccetpw5zem\Url.2txcxmkov3cbmd5tmqeaz0ccetpw5zem\Files\AnotherTopLevelDirectory\NewFile.txt'.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   --- End of inner exception stack trace ---
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   at System.IO.IsolatedStorage.IsolatedStorageFile.CreateFile(String path)
   at tempcode.Program.Main() in C:\temp\tempcode\Program.cs:line 18
Created directories.
Created a new file in the root.
Unhandled exception. System.IO.IsolatedStorage.IsolatedStorageException: Operation not permitted on IsolatedStorageFileStream.
 ---> System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\thrak\AppData\Local\IsolatedStorage\5uysulys.woc\5ou1xfgf.iws\Url.2txcxmkov3cbmd5tmqeaz0ccetpw5zem\Url.2txcxmkov3cbmd5tmqeaz0ccetpw5zem\Files\AnotherTopLevelDirectory\InsideDirectory\HereIAm.txt'.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   --- End of inner exception stack trace ---
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   at System.IO.IsolatedStorage.IsolatedStorageFile.CreateFile(String path)
   at tempcode.Program.Main() in C:\temp\tempcode\Program.cs:line 35
```

adegeo reviewed (changes requested):
Looks like there may have been an error in one of the commits. Asking collaborator for feedback.

BillWagner commented:
@OzBob This is getting stale. Can you respond to the previous comment?

adegeo commented:
@OzBob Going to close this soon as abandoned. Just FYI 😄 Please respond, cheers!

OzBob reviewed (commented):
what error is holding this up? 
can you suggest how to resolve?

OzBob commented:
what does 'project missing' mean in https://github.com/dotnet/docs/pull/18769/checks?check_run_id=994903800?
and how do you suggest it be resolved?

adegeo reviewed (changes requested):
Code doesn't seem to work as submitted. Please see inline comment.

adegeo commented:
> what does 'project missing' mean in https://github.com/dotnet/docs/pull/18769/checks?check_run_id=994903800?
> and how do you suggest it be resolved?

This error isn't important for the PR and can be ignored. We've moved to a different snippet processing system and this snippet is part of the old system. We're not requiring community members to create projects (required by the new system) when they are doing updates to old snippets.

OzBob commented:
> We're not requiring community members to create projects
https://github.com/dotnet/docs/pull/18769/checks?check_run_id=994903800
still says:
'missing project'

> Code doesn't seem to work as submitted. Please see inline comment.
![image](https://user-images.githubusercontent.com/1796771/95944202-438a2e80-0e1a-11eb-9dbe-80b51491c22c.png)

I have a try/catch to prove a point, but the 2nd exception in the inline comment can't be replicated.

My last line of code is to write "Created a new file in the root." to the COnsole. There is no code after that. What line of code in your copy of source.cs is causing that exception?

adegeo commented:
Sorry for the delay. This notification slipped through. Try deleting your isolated storage folder to clear it out.

Using the code from your commit: https://github.com/dotnet/docs/blob/bb8865e133e5751407e4eb3ecebf1c4c9e663272/samples/snippets/csharp/VS_Snippets_CLR/conceptual.isolatedstorage/cs/source.cs

I get this error:

```
C:\code\Work\dotnet\test\a> dotnet run
System.IO.IsolatedStorage.IsolatedStorageException: Operation not permitted on IsolatedStorageFileStream.
 ---> System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\thrak\AppData\Local\IsolatedStorage\5uysulys.woc\5ou1xfgf.iws\Url.3xes0z4spj2hngysuwp1t4k0om5g1pbk\Url.3xes0z4spj2hngysuwp1t4k0om5g1pbk\Files\AnotherTopLevelDirectory\NewFile.txt'.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   --- End of inner exception stack trace ---
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   at System.IO.IsolatedStorage.IsolatedStorageFile.CreateFile(String path)
   at a.Program.Main(String[] args) in C:\code\Work\dotnet\test\a\Program.cs:line 18
Created directories.
Created a new file in the root.
Unhandled exception. System.IO.IsolatedStorage.IsolatedStorageException: Operation not permitted on IsolatedStorageFileStream.
 ---> System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\thrak\AppData\Local\IsolatedStorage\5uysulys.woc\5ou1xfgf.iws\Url.3xes0z4spj2hngysuwp1t4k0om5g1pbk\Url.3xes0z4spj2hngysuwp1t4k0om5g1pbk\Files\AnotherTopLevelDirectory\InsideDirectory\HereIAm.txt'.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   --- End of inner exception stack trace ---
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   at System.IO.IsolatedStorage.IsolatedStorageFile.CreateFile(String path)
   at a.Program.Main(String[] args) in C:\code\Work\dotnet\test\a\Program.cs:line 35
```

If I run it a second time, I get this error:

```
C:\code\Work\dotnet\test\a> dotnet run
Unhandled exception. System.IO.IsolatedStorage.IsolatedStorageException: Operation not permitted on IsolatedStorageFileStream.
 ---> System.IO.IOException: The process cannot access the file 'C:\Users\thrak\AppData\Local\IsolatedStorage\5uysulys.woc\5ou1xfgf.iws\Url.3xes0z4spj2hngysuwp1t4k0om5g1pbk\Url.3xes0z4spj2hngysuwp1t4k0om5g1pbk\Files\AnotherTopLevelDirectory\NewFile.txt' because it is being used by another process.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   --- End of inner exception stack trace ---
   at System.IO.IsolatedStorage.IsolatedStorageFileStream.InitializeFileStream(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, IsolatedStorageFile isf)
   at System.IO.IsolatedStorage.IsolatedStorageFile.CreateFile(String path)
   at a.Program.Main(String[] args) in C:\code\Work\dotnet\test\a\Program.cs:line 27
```

Here is the code I used.


```csharp
using System;
using System.IO;
using System.IO.IsolatedStorage;

namespace a
{
    class Program
    {
        static void Main(string[] args)
        {
            using (IsolatedStorageFile isoStore = IsolatedStorageFile.GetStore(IsolatedStorageScope.User | IsolatedStorageScope.Domain | IsolatedStorageScope.Assembly, null, null))
            {
                isoStore.CreateDirectory("TopLevelDirectory");
                isoStore.CreateDirectory("TopLevelDirectory/SecondLevel");
                var isofilename = "AnotherTopLevelDirectory/NewFile.txt";
                try
                {
                    isoStore.CreateFile(isofilename);
                }
                catch (IsolatedStorageException iex)
                {
                    Console.WriteLine(iex.ToString());
                }

                var path = Path.GetDirectoryName(isofilename);
                isoStore.CreateDirectory(path);//Creating "AnotherTopLevelDirectory";
                isoStore.CreateFile(isofilename);//Creating "AnotherTopLevelDirectory/NewFile.txt";

                isoStore.CreateFile("AnotherTopLevelDirectory/InsideDirectory");
                Console.WriteLine("Created directories.");

                isoStore.CreateFile("InTheRoot.txt");
                Console.WriteLine("Created a new file in the root.");

                isoStore.CreateFile("AnotherTopLevelDirectory/InsideDirectory/HereIAm.txt");
                Console.WriteLine("Created a new file in the InsideDirectory.");
            }
        }
    }
}
```

OzBob commented:
investigating ...

adegeo commented:
Closing this as it's gone stale. Feel free to reopen.
